### PR TITLE
added spi, timer, and (WIP) serial peripherals

### PR DESCRIPTION
**Pull request by o080o**
* implemented FullDuplex SPI. tested both SPI0 and SPI1 with real devices :D
* implemented a single Timer as a CountDown timer, to do more accurate delay_ms stuff than busy-looping :D (Timer6, the "basic" timer). still need to make it a macro/generic to work with timers0-5.
* very WIP USART code, that lets me at least print things to a console without a debugger :'(
* added a Clocks struct so that I could abstract the clock frequency calculation in the SPI and Timer objects, but currently only hardcodes the default clock values, and doesn't indicate that the clock speeds are frozen in any way.

### CODE REVIEW

**Comment by luojia65** [comment]
Thanks! I'll have these modules tested soon. Merge now :)

File: src/rcu.rs [hunk 1]
```diff
@@ -78,7 +94,64 @@ impl APB2 {
...
+pub struct AHB {
+    _ownership: ()
 }
 
+impl AHB {
+    #[inline]
+    pub(crate) fn en(&mut self) -> &rcu::AHBEN {
+        unsafe { &(*RCU::ptr()).ahben}
+    }
+}
```
Comment (luojia65) [inline src/rcu.rs:157]: Note: `rst()`

File: src/gpio.rs [hunk 2]
```diff
@@ -239,7 +239,7 @@ pub mod $gpiox {
 
         fn split(self, apb2: &mut APB2) -> Self::Parts {
             riscv::interrupt::free(|_| {
-                apb2.en().write(|w| w.$en().set_bit());
+                apb2.en().modify(|_,w| w.$en().set_bit());
```
Comment (luojia65) [inline src/gpio.rs:242]: Why should we perform this change? :) @o080o

File: src/timer.rs [hunk 3]
```diff
@@ -1,46 +1,71 @@
...
 use crate::rcu;
-use crate::pac::TIMER2;
+use crate::pac::TIMER6;
 
-impl Timer<TIMER2> {
-    pub fn timer2<T>(timer2: TIMER2, timeout: T, clocks: rcu::Clocks, apb1: &mut rcu::APB1) -> Self 
-    where 
-        T: Into<Hertz>
+impl Timer6 {
+    pub fn new(timer: TIMER6, clock: rcu::Clocks, apb1: &mut rcu::APB1) -> Self
```
Comment (luojia65) [inline src/timer.rs:18]: Could it be better if we implement for `Timer<pac::TIM6>` other than `Timer6`?

File: Cargo.toml [hunk 4]
```diff
@@ -16,6 +16,10 @@ embedded-hal = { version = "0.2.3", features = ["unproven"] }
 nb = "0.1"
 riscv = "0.5"
 
+[dependencies.void]
+default-features = false
+version = "1.0.2"
```
Comment (luojia65) [inline Cargo.toml:21]: Maybe we could use `Infallible` in `core::convert` by now, it does not require another dependency.

File: src/gpio.rs (hunk 2 above)
Comment (o080o) [reply src/gpio.rs:242]: without this, I was unable to enable more than one gpio port. It seems that write resets all other values, and modify does not.

**Comment by torntrousers** [comment]
This looks great. Any chance you could give a code example of how to use the timer and uart to write debug messages?

**Pull request merged**